pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -1117,3 +1117,348 @@ where
+					to_pool(done_amount)
+				}
+				Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+					done_amount, ..
+				} => to_pool(done_amount),
+				Self::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+					done_amount,
+					..
+				} => to_pool(done_amount),
+				Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+					done_amount,
+					..
+				} => to_pool(done_amount),
+				Self::SwapIntoForeignDone { done_swap } => to_pool(done_swap.amount),
+				Self::SwapIntoForeignDoneAndInvestmentOngoing { done_swap, .. } => {
+					to_pool(done_swap.amount)
+				}
+				_ => 0,
+			}
+		}
+
+		fn get_swap_pool_amount(&self) -> u128 {
+			self.get_active_swap_amount_pool_denominated().unwrap()
+		}
+
+		fn has_active_pool_swap(&self) -> bool {
+			self.get_active_swap()
+				.map(|swap| swap.currency_in == CurrencyId::Pool)
+				.unwrap_or(false)
+		}
+
+		fn total(&self) -> u128 {
+			self.get_investing_amount() + self.get_done_amount() + self.get_swap_pool_amount()
+		}
+	}
+
+	struct Checker {
+		old_state: InvestState,
+	}
+
+	impl Checker {
+		fn new(initial_state: InvestState, use_case: &[InvestTransition]) -> Self {
+			println!("Testing use case: {:#?}", use_case);
+
+			Self {
+				old_state: initial_state,
+			}
+		}
+
+		/// Invariants from: https://centrifuge.hackmd.io/IPtRlOrOSrOF9MHjEY48BA?view#Without-storage
+		fn check_delta_invariant(&self, transition: &InvestTransition, new_state: &InvestState) {
+			println!("Transition: {:#?}", transition);
+			println!("New state: {:#?}, total: {}", new_state, new_state.total());
+
+			match *transition {
+				InvestTransition::IncreaseInvestOrder(swap) => {
+					let diff = new_state.total() - self.old_state.total();
+					assert_eq!(diff, swap.amount);
+				}
+				InvestTransition::DecreaseInvestOrder(_) => {
+					let diff = new_state.total() - self.old_state.total();
+					assert_eq!(diff, 0);
+				}
+				InvestTransition::FulfillSwapOrder(swap) => {
+					let diff = new_state.total() - self.old_state.total();
+					assert_eq!(diff, 0);
+
+					if self.old_state.has_active_pool_swap() {
+						let invest_diff = new_state.get_investing_amount()
+							- self.old_state.get_investing_amount();
+						assert_eq!(invest_diff, swap.amount)
+					} else {
+						let done_diff =
+							new_state.get_done_amount() - self.old_state.get_done_amount();
+						assert_eq!(done_diff, to_pool(swap.amount))
+					}
+				}
+				InvestTransition::CollectInvestment(value) => {
+					if self.old_state.get_investing_amount() == 0 {
+						assert_eq!(new_state.get_investing_amount(), 0)
+					} else {
+						assert_eq!(new_state.get_investing_amount(), value);
+
+						assert_eq!(
+							new_state.get_done_amount(),
+							self.old_state.get_done_amount()
+						);
+						assert_eq!(
+							new_state.get_swap_pool_amount(),
+							self.old_state.get_swap_pool_amount(),
+						);
+					}
+				}
+			}
+		}
+	}
+
+	#[test]
+	fn fuzzer() {
```

Comment:
👌

Reply:
Not sure if it should be fuzzier by the way haha 🤔

Reply:
Confirmed, it's fuzzer :)